src/dataflow/operators/join_operator.rs

```diff
@@ -1,83 +1,179 @@
+use crate::dataflow::message::Message;
+use crate::dataflow::stream::WriteStreamT;
+use crate::dataflow::{Data, Operator, OperatorConfig, ReadStream, Timestamp, WriteStream};
+use chashmap::CHashMap;
+use serde::Deserialize;
+use std::cmp::Reverse;
+use std::collections::BinaryHeap;
 use std::marker::PhantomData;
+use std::sync::{Arc, RwLock};
 
-use crate::{
-    add_watermark_callback,
-    dataflow::{Data, Operator, OperatorConfig, ReadStream, Timestamp, WriteStream},
-    OperatorId,
-};
-
-#[derive(Clone, Default)]
-struct WindowState<D> {
-    pub msgs: Vec<D>,
+/// A structure that stores the state associated with a stream for the JoinOperator, and provides
+/// the associated functions for mutation of the data.
+/// Uses a ConcurrentHashMap to store the messages and a min-heap to ensure easy retrieval of the
+/// timestamps for cleaning.
+#[derive(Clone)]
+struct StreamState<D: Data> {
+    // We use a CHashMap because multiple callbacks can be simultaneously invoked.
+    msgs: Arc<RwLock<CHashMap<Timestamp, Vec<D>>>>,
+    // A min-heap tracking the keys of the hashmap.
+    _timestamps: Arc<RwLock<BinaryHeap<Reverse<Timestamp>>>>,
 }
 
-impl<D> WindowState<D> {
-    pub fn new() -> Self {
-        Self { msgs: Vec::new() }
+impl<D: Data> StreamState<D> {
+    fn new() -> Self {
+        Self {
+            msgs: Arc::new(RwLock::new(CHashMap::new())),
+            _timestamps: Arc::new(RwLock::new(BinaryHeap::new())),
+        }
+    }
+
+    /// Adds a message to the ConcurrentHashMap.
+    fn add_msg(&mut self, timestamp: &Timestamp, msg: D) {
+        // Insert a new Vector if the key does not exist, and add the key to the timestamps.
+        let _msgs = &(*(*self.msgs).write().unwrap());
+        if !_msgs.contains_key(timestamp) {
+            _msgs.insert_new(timestamp.clone(), Vec::new());
+            self._timestamps
+                .write()
+                .unwrap()
+                .push(Reverse(timestamp.clone()));
+        }
+
+        // Get the vector corresponding to the given key and acquire a write lock on it.
+        let mut messages = _msgs.get_mut(timestamp).unwrap();
+        messages.push(msg);
+        // The data structure automatically releases the write lock once the guard goes out of
+        // scope.
+    }
+
+    /// Cleans the state corresponding to a given Timestamp (upto and including).
+    fn clean_state(&self, timestamp: &Timestamp) {
+        let _timestamps = &mut (*(*self._timestamps).write().unwrap());
+        while _timestamps.len() > 0 && _timestamps.peek().unwrap().0 <= *timestamp {
```

Comment:
Might need some small fixes but this  should work:
```suggestion
        while timestamps.peek().map_or(false, |t| t <= timestamp) {
```